File: LeetCode/medium/Find_All_Anagrams_in_a_String.cxx
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> a(26, 0);
        vector<int> c(26, 0);

        for (auto x : p)
            a[x-'a']++;

        if (p.size() > s.size())
            return {};

        for (int i = 0; i < p.size(); ++i) {
            c[s[i]-'a']++;
        }
        vector<int> res;
        for (int i = 0; i < s.size()-p.size(); ++i) {
            if (a == c)
                res.push_back(i);
            c[s[i]-'a']--;
            c[s[i+p.size()]-'a']++;
        }
        if (a == c)
            res.push_back(s.size()-p.size());
        return res;
    }
};

File: LeetCode/medium/Find_All_Anagrams_in_a_String.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        letters = [0] * 26
        ana = [0] * 26

        for c in p:
            ana[ord(c)-ord('a')] += 1

        if len(s) < len(p):
            return []
        res = []
        
        l = 0
        for i in range(len(p)):
            letters[ord(s[i])-ord('a')] += 1
        while l+len(p) <= len(s):
            if letters == ana:
                res.append(l)
            letters[ord(s[l])-ord('a')] -= 1
            if l+len(p) < len(s):
                letters[ord(s[l+len(p)])-ord('a')] += 1
            l += 1
        
        return res
